Feature: unify /create keys with /transact

cap10morgan opened the pull request:
Instead of `ledger` & `txn` use `@id` & `@graph` like `/transact` does. Checks that the ledger named by `@id` doesn't yet exist, creates it if not, and then transacts the data in `@graph` into it.

I was working on syncing up the latest changes in here w/ server and noticed this inconsistency in both code bases. Not sure if we intentionally left it like this just to change things incrementally (which is good), but it seemed good either way to make this consistent.

mpoffald commented on src/fluree/http_api/handlers/ledger.clj:
```diff
@@ -65,19 +65,24 @@
-          (let [opts*    (cond-> (opts->context-type opts content-type)
-                           defaultContext (assoc :defaultContext defaultContext)
-                           did (assoc :did did))
+          (let [default-context (get body "@context")
```
This isn't quite in line with what happens for `/transact`. For transactions, the top-level context is interpreted as the context for the transaction document itself, as well as for the data contained within the graph (which will be combined with any underlying default-context as applicable).

We use an iri for default context (defined [here](https://github.com/fluree/db/blob/main/src/fluree/db/constants.cljc#L36)).

cap10morgan replied:
Thanks! I had misinterpreted that. Will fix.

mpoffald commented on src/fluree/http_api/components/http.clj:
```diff
@@ -40,13 +41,20 @@
+    (cond-> {"@id"    (:id req)
+             "@graph" (:graph req)}
+            (contains? req :context) (assoc "@context" (:context req)))
+    req))
```
If I'm understanding this correctly, this means users can't use an aliased `@id` or `@graph` in the `/create` endpoint, right?

This is different from `/transact`, which expands top-level keys using any provided `@context` before looking for the id and graph keys: https://github.com/fluree/db/blob/main/src/fluree/db/api/transact.cljc#L33.  So those keys can be aliased in the top-level context if desired. All of that parsing work is done on the `db` side.

mpoffald reviewed (commented):
>  Not sure if we intentionally left it like this just to change things incrementally

As the person who did the `/transact` update, I can tell you that I just didn't update this because it wasn't in the ticket, and I wasn't sure what the plans were. 😄 

But those changes to `/transact` introduced the idea of operating on a conn and doing ledger lookups. This makes the `transact!` process on the db side look _extremely_ similar to the `/create` process, only we don't currently allow `transact!` to create ledgers. We discussed a bit whether it should, but afaict that discussion did not reach a conclusion, so I didn't include it in the `/transact` work.

Tbh though, if we want `/create` to handle keys exactly like `/transact` currently does, we'd want to do the top-level expansion work that db's `transact!` does, so that aliases are allowed. Or, we could just let `transact!` create ledgers (maybe using some opt so it doesn't always happen?).

cap10morgan commented:
> > Not sure if we intentionally left it like this just to change things incrementally
> 
> As the person who did the `/transact` update, I can tell you that I just didn't update this because it wasn't in the ticket, and I wasn't sure what the plans were. 😄
> 
> But those changes to `/transact` introduced the idea of operating on a conn and doing ledger lookups. This makes the `transact!` process on the db side look _extremely_ similar to the `/create` process, only we don't currently allow `transact!` to create ledgers. We discussed a bit whether it should, but afaict that discussion did not reach a conclusion, so I didn't include it in the `/transact` work.
> 
> Tbh though, if we want `/create` to handle keys exactly like `/transact` currently does, we'd want to do the top-level expansion work that db's `transact!` does, so that aliases are allowed. Or, we could just let `transact!` create ledgers (maybe using some opt so it doesn't always happen?).

Thanks! This is all very useful. I think I'd vote for keeping `/create` as a distinct endpoint / op based on the Nexus' teams early feedback that our original "create is a mode of transact" implementation was counterintuitive and confusing.

I'll make the key handling vis-a-vis context correct along the lines of transact. Thanks for catching that (and sorry I didn't see where transact was doing it)!

cap10morgan commented:
UPDATE to address earlier feedback:

I made this work more like `/transact` and fixed the shortcomings @mpoffald pointed out in the original implementation. In order to be more consistent w/ `/transact`, I moved some of the functionality down into db lib. The PR for that is here: https://github.com/fluree/db/pull/570

I also converted `error-catching-handler` into a macro named `defhandler` b/c all of the errors were reporting that they came from the `error-catching-handler` fn instead of the actual handler code.

mpoffald commented on test/fluree/http_api/integration/basic_transaction_test.clj:
```diff
@@ -11,11 +11,11 @@
-                                           "type"    "foo:test"
-                                           "ex:name" "create-endpoint-test"}]})
+                       {"@id"      ledger-name
+                        "@context" ["" {"foo" "http://foobar.com/"}]
```
I think I may be confused, I expected these `defaultContext`s to change to the [default context iri](https://github.com/fluree/db/blob/main/src/fluree/db/constants.cljc#L36), rather than to `@context`

cap10morgan replied:
It should work either way in this case, right? I did change one of them but I was thinking it would be good to make sure these got merged into the transaction context too. But if it's better to use the defaultContext IRI here too I'd be happy to.

mpoffald replied:
I actually missed that you did change one of them, I see that now. 😄  I had thought that they all became `@context`s, and that seemed potentially Meaningful, so I wanted to double-check my understanding. But no, it doesn't matter for what these tests are actually testing.

mpoffald reviewed (approved):
🥪
